File: android/app/src/main/java/com/irrgenius/android/data/models/SavedCalculation.kt
package com.irrgenius.android.data.models

import androidx.room.Entity
import androidx.room.PrimaryKey
import java.time.LocalDateTime
import java.util.UUID

// Import the FollowOnInvestmentValidationException from FollowOnInvestment.kt
// This will be available once both files are in the same package

// Validation exception for SavedCalculation
class SavedCalculationValidationException(message: String) : Exception(message)

@Entity(tableName = "saved_calculations")
data class SavedCalculation(
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val name: String,
    val calculationType: CalculationMode,
    val createdDate: LocalDateTime,
    val modifiedDate: LocalDateTime,
    val projectId: String?,
    
    // Calculation inputs
    val initialInvestment: Double?,
    val outcomeAmount: Double?,
    val timeInMonths: Double?,
    val irr: Double?,
    
    // Portfolio Unit Investment specific parameters
    val unitPrice: Double?,
    val successRate: Double?, // Percentage (0-100)
    val outcomePerUnit: Double?,
    val investorShare: Double?, // Percentage (0-100)
    val feePercentage: Double?, // Management/legal/servicing fees percentage (0-100)
    
    // Results
    val calculatedResult: Double?,
    val growthPointsJson: String?, // JSON serialized GrowthPoint array
    
    // Metadata
    val notes: String?,
    val tags: String? // JSON array as string
) {
    
    companion object {
        private const val MAX_NAME_LENGTH = 100
        private const val MIN_IRR = -100.0
        private const val MAX_IRR = 1000.0
        
        /**
         * Validates calculation name
         */
        @Throws(SavedCalculationValidationException::class)
        fun validateName(name: String) {
            if (name.trim().isEmpty()) {
                throw SavedCalculationValidationException("Calculation name cannot be empty")
            }
            
            if (name.length > MAX_NAME_LENGTH) {
                throw SavedCalculationValidationException("Name too long (max $MAX_NAME_LENGTH characters)")
            }
            
            // Check for invalid characters
            val invalidChars = listOf('<', '>', ':', '"', '/', '\\', '|', '?', '*')
            if (name.any { it in invalidChars }) {
                throw SavedCalculationValidationException("Name contains invalid characters")
            }
        }
        
        /**
         * Validates calculation inputs based on calculation type
         */
        @Throws(SavedCalculationValidationException::class)
        fun validateCalculationInputs(
            calculationType: CalculationMode,
            initialInvestment: Double?,
            outcomeAmount: Double?,
            timeInMonths: Double?,
            irr: Double?,
            unitPrice: Double? = null,
            successRate: Double? = null,
            outcomePerUnit: Double? = null,
            investorShare: Double? = null
        ) {
            val missingFields = mutableListOf<String>()
            
            when (calculationType) {
                CalculationMode.CALCULATE_IRR -> {
                    initialInvestment?.let { investment ->
                        if (investment <= 0) throw SavedCalculationValidationException("Initial investment must be positive")
                    } ?: missingFields.add("Initial Investment")
                    
                    outcomeAmount?.let { outcome ->
                        if (outcome <= 0) throw SavedCalculationValidationException("Outcome amount must be positive")
                    } ?: missingFields.add("Outcome Amount")
                    
                    timeInMonths?.let { time ->
                        if (time <= 0) throw SavedCalculationValidationException("Time in months must be positive")
                    } ?: missingFields.add("Time in Months")
                }
                
                CalculationMode.CALCULATE_OUTCOME -> {
                    initialInvestment?.let { investment ->
                        if (investment <= 0) throw SavedCalculationValidationException("Initial investment must be positive")
                    } ?: missingFields.add("Initial Investment")
                    
                    irr?.let { irrValue ->
                        if (irrValue <= MIN_IRR || irrValue >= MAX_IRR) {
                            throw SavedCalculationValidationException("IRR must be between $MIN_IRR and $MAX_IRR")
                        }
                    } ?: missingFields.add("IRR")
                    
                    timeInMonths?.let { time ->
                        if (time <= 0) throw SavedCalculationValidationException("Time in months must be positive")
                    } ?: missingFields.add("Time in Months")
                }
                
                CalculationMode.CALCULATE_INITIAL -> {
                    outcomeAmount?.let { outcome ->
                        if (outcome <= 0) throw SavedCalculationValidationException("Outcome amount must be positive")
                    } ?: missingFields.add("Outcome Amount")
                    
                    irr?.let { irrValue ->
                        if (irrValue <= MIN_IRR || irrValue >= MAX_IRR) {
                            throw SavedCalculationValidationException("IRR must be between $MIN_IRR and $MAX_IRR")
                        }
                    } ?: missingFields.add("IRR")
                    
                    timeInMonths?.let { time ->
                        if (time <= 0) throw SavedCalculationValidationException("Time in months must be positive")
                    } ?: missingFields.add("Time in Months")
                }
                
                CalculationMode.CALCULATE_BLENDED -> {
                    initialInvestment?.let { investment ->
                        if (investment <= 0) throw SavedCalculationValidationException("Initial investment must be positive")
                    } ?: missingFields.add("Initial Investment")
                    
                    outcomeAmount?.let { outcome ->
                        if (outcome <= 0) throw SavedCalculationValidationException("Outcome amount must be positive")
                    } ?: missingFields.add("Outcome Amount")
                    
                    timeInMonths?.let { time ->
                        if (time <= 0) throw SavedCalculationValidationException("Time in months must be positive")
                    } ?: missingFields.add("Time in Months")
                }
                
                CalculationMode.PORTFOLIO_UNIT_INVESTMENT -> {
                    initialInvestment?.let { investment ->
                        if (investment <= 0) throw SavedCalculationValidationException("Initial investment must be positive")
                    } ?: missingFields.add("Investment Amount")
                    
                    unitPrice?.let { price ->
                        if (price <= 0) throw SavedCalculationValidationException("Unit price must be positive")
                    } ?: missingFields.add("Unit Price")
                    
                    successRate?.let { rate ->
                        if (rate < 0 || rate > 100) throw SavedCalculationValidationException("Success rate must be between 0 and 100")
                    } ?: missingFields.add("Success Rate")
                    
                    outcomePerUnit?.let { outcome ->
                        if (outcome <= 0) throw SavedCalculationValidationException("Outcome per unit must be positive")
                    } ?: missingFields.add("Outcome Per Unit")
                    
                    investorShare?.let { share ->
                        if (share < 0 || share > 100) throw SavedCalculationValidationException("Investor share must be between 0 and 100")
                    } ?: missingFields.add("Investor Share")
                    
                    timeInMonths?.let { time ->
                        if (time <= 0) throw SavedCalculationValidationException("Time in months must be positive")
                    } ?: missingFields.add("Time in Months")
                }
            }
            
            if (missingFields.isNotEmpty()) {
                throw SavedCalculationValidationException("Missing required fields: ${missingFields.joinToString(", ")}")
            }
        }
        
        /**
         * Creates a validated SavedCalculation instance
         */
        @Throws(SavedCalculationValidationException::class)
        fun createValidated(
            id: String = UUID.randomUUID().toString(),
            name: String,
            calculationType: CalculationMode,
            createdDate: LocalDateTime = LocalDateTime.now(),
            modifiedDate: LocalDateTime = LocalDateTime.now(),
            projectId: String? = null,
            initialInvestment: Double? = null,
            outcomeAmount: Double? = null,
            timeInMonths: Double? = null,
            irr: Double? = null,
            unitPrice: Double? = null,
            successRate: Double? = null,
            outcomePerUnit: Double? = null,
            investorShare: Double? = null,
            feePercentage: Double? = null,
            calculatedResult: Double? = null,
            growthPointsJson: String? = null,
            notes: String? = null,
            tags: String? = null
        ): SavedCalculation {
            validateName(name)
            validateCalculationInputs(calculationType, initialInvestment, outcomeAmount, timeInMonths, irr, unitPrice, successRate, outcomePerUnit, investorShare)
            
            return SavedCalculation(
                id = id,
                name = name,
                calculationType = calculationType,
                createdDate = createdDate,
                modifiedDate = modifiedDate,
                projectId = projectId,
                initialInvestment = initialInvestment,
                outcomeAmount = outcomeAmount,
                timeInMonths = timeInMonths,
                irr = irr,
                unitPrice = unitPrice,
                successRate = successRate,
                outcomePerUnit = outcomePerUnit,
                investorShare = investorShare,
                feePercentage = feePercentage,
                calculatedResult = calculatedResult,
                growthPointsJson = growthPointsJson,
                notes = notes,
                tags = tags
            )
        }
    }
    
    /**
     * Validates the entire calculation for data integrity
     */
    @Throws(SavedCalculationValidationException::class)
    fun validate() {
        validateName(name)
        validateCalculationInputs(calculationType, initialInvestment, outcomeAmount, timeInMonths, irr, unitPrice, successRate, outcomePerUnit, investorShare)
    }
    
    /**
     * Checks if calculation has all required data for its type
     */
    val isComplete: Boolean
        get() = try {
            validate()
            true
        } catch (e: SavedCalculationValidationException) {
            false
        }
    
    /**
     * Returns a summary of the calculation for display purposes
     */
    val summary: String
        get() = when (calculationType) {
            CalculationMode.CALCULATE_IRR -> {
                calculatedResult?.let { result ->
                    "IRR: ${String.format("%.2f", result)}%"
                } ?: "IRR calculation"
            }
            CalculationMode.CALCULATE_OUTCOME -> {
                calculatedResult?.let { result ->
                    "Outcome: $${String.format("%.2f", result)}"
                } ?: "Outcome calculation"
            }
            CalculationMode.CALCULATE_INITIAL -> {
                calculatedResult?.let { result ->
                    "Initial: $${String.format("%.2f", result)}"
                } ?: "Initial investment calculation"
            }
            CalculationMode.CALCULATE_BLENDED -> {
                calculatedResult?.let { result ->
                    "Blended IRR: ${String.format("%.2f", result)}%"
                } ?: "Blended IRR calculation"
            }
            CalculationMode.PORTFOLIO_UNIT_INVESTMENT -> {
                calculatedResult?.let { result ->
                    "Portfolio IRR: ${String.format("%.2f", result)}%"
                } ?: "Portfolio unit investment"
            }
        }
    
    /**
     * Creates a copy with updated modification date
     */
    fun withUpdatedModificationDate(): SavedCalculation {
        return copy(modifiedDate = LocalDateTime.now())
    }
    
    /**
     * Serializes tags array to JSON string
     */
    fun getTagsAsJson(): String {
        return tags ?: "[]"
    }
    
    /**
     * Deserializes tags from JSON string
     */
    fun getTagsFromJson(): List<String> {
        return try {
            if (tags.isNullOrEmpty()) return emptyList()
            // Simple JSON parsing for string arrays
            val cleanJson = tags.trim()
            if (cleanJson.startsWith("[") && cleanJson.endsWith("]")) {
                val content = cleanJson.substring(1, cleanJson.length - 1)
                if (content.isBlank()) return emptyList()
                return content.split(",").map { it.trim().removeSurrounding("\"") }
            }
            emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * Gets growth points from JSON string
     */
    fun getGrowthPoints(): List<GrowthPoint> {
        return GrowthPoint.fromJsonString(growthPointsJson)
    }
    
    /**
     * Sets growth points as JSON string
     */
    fun withGrowthPoints(growthPoints: List<GrowthPoint>): SavedCalculation {
        return copy(growthPointsJson = GrowthPoint.toJsonString(growthPoints))
    }
    
    /**
     * Sets tags as JSON string
     */
    fun withTags(tagsList: List<String>): SavedCalculation {
        val tagsJson = if (tagsList.isEmpty()) {
            "[]"
        } else {
            "[" + tagsList.joinToString(",") { "\"$it\"" } + "]"
        }
        return copy(tags = tagsJson)
    }
}


@Entity(
    tableName = "follow_on_investments",
    foreignKeys = [androidx.room.ForeignKey(
        entity = SavedCalculation::class,
        parentColumns = ["id"],
        childColumns = ["calculationId"],
        onDelete = androidx.room.ForeignKey.CASCADE
    )]
)
data class FollowOnInvestmentEntity(
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val calculationId: String,
    val amount: Double,
    val investmentType: InvestmentType,
    val timingType: TimingType,
    val absoluteDate: String?, // ISO date string
    val relativeTime: Double?,
    val relativeTimeUnit: TimeUnit?,
    val valuationMode: ValuationMode,
    val valuationType: ValuationType,
    val customValuation: Double?,
    val irr: Double? // Used for computed valuation
) {
    
    /**
     * Converts this FollowOnInvestmentEntity to a FollowOnInvestment domain model
     */
    fun toDomainModel(): FollowOnInvestment {
        return FollowOnInvestment(
            id = id,
            amount = amount,
            investmentType = investmentType,
            timingType = timingType,
            absoluteDate = absoluteDate?.let { java.time.LocalDate.parse(it) } ?: java.time.LocalDate.now(),
            relativeTime = relativeTime ?: 1.0,
            relativeTimeUnit = relativeTimeUnit ?: TimeUnit.YEARS,
            valuationMode = valuationMode,
            valuationType = valuationType,
            customValuation = customValuation ?: 0.0,
            irr = irr ?: 0.0
        )
    }
    
    /**
     * Validates the entity data
     */
    @Throws(FollowOnInvestmentValidationException::class)
    fun validate() {
        if (amount <= 0) {
            throw FollowOnInvestmentValidationException("Investment amount must be positive")
        }
        
        when (timingType) {
            TimingType.ABSOLUTE -> {
                if (absoluteDate == null) {
                    throw FollowOnInvestmentValidationException("Absolute date is required for absolute timing")
                }
                try {
                    java.time.LocalDate.parse(absoluteDate)
                } catch (e: Exception) {
                    throw FollowOnInvestmentValidationException("Invalid date format")
                }
            }
            TimingType.RELATIVE -> {
                if (relativeTime == null || relativeTime <= 0) {
                    throw FollowOnInvestmentValidationException("Relative time must be positive")
                }
                if (relativeTimeUnit == null) {
                    throw FollowOnInvestmentValidationException("Relative time unit is required")
                }
            }
        }
        
        when (valuationMode) {
            ValuationMode.CUSTOM -> {
                if (valuationType == ValuationType.SPECIFIED) {
                    if (customValuation == null || customValuation <= 0) {
                        throw FollowOnInvestmentValidationException("Custom valuation must be positive")
                    }
                } else {
                    if (irr == null || irr <= -100.0 || irr >= 1000.0) {
                        throw FollowOnInvestmentValidationException("IRR must be between -100 and 1000")
                    }
                }
            }
            ValuationMode.TAG_ALONG -> {
                // Tag-along mode doesn't require additional validation
            }
        }
    }
    
    /**
     * Checks if the entity has valid data
     */
    val isValid: Boolean
        get() = try {
            validate()
            true
        } catch (e: FollowOnInvestmentValidationException) {
            false
        }
}

File: ios/IRR Genius/Models/SavedCalculation.swift
//
//  SavedCalculation.swift
//  IRR Genius
//
//

import Foundation

// MARK: - Validation Errors
enum SavedCalculationValidationError: LocalizedError {
    case emptyName
    case invalidName(String)
    case negativeInvestment
    case negativeOutcome
    case invalidTimeInMonths
    case invalidIRR
    case invalidFollowOnInvestments([String])
    case missingRequiredFields([String])
    
    var errorDescription: String? {
        switch self {
        case .emptyName:
            return "Calculation name cannot be empty"
        case .invalidName(let reason):
            return "Invalid calculation name: \(reason)"
        case .negativeInvestment:
            return "Initial investment must be positive"
        case .negativeOutcome:
            return "Outcome amount must be positive"
        case .invalidTimeInMonths:
            return "Time in months must be positive"
        case .invalidIRR:
            return "IRR must be a valid percentage"
        case .invalidFollowOnInvestments(let errors):
            return "Invalid follow-on investments: \(errors.joined(separator: ", "))"
        case .missingRequiredFields(let fields):
            return "Missing required fields: \(fields.joined(separator: ", "))"
        }
    }
}

struct SavedCalculation: Identifiable, Codable {
    let id: UUID
    let name: String
    let calculationType: CalculationMode
    let createdDate: Date
    let modifiedDate: Date
    let projectId: UUID?
    
    // Calculation inputs (varies by type)
    let initialInvestment: Double?
    let outcomeAmount: Double?
    let timeInMonths: Double?
    let irr: Double?
    let followOnInvestments: [FollowOnInvestment]?
    
    // Portfolio Unit Investment specific parameters
    let unitPrice: Double?
    let successRate: Double? // Percentage (0-100)
    let outcomePerUnit: Double?
    let investorShare: Double? // Percentage (0-100)
    let feePercentage: Double? // Management/legal/servicing fees percentage (0-100)
    
    // Results
    let calculatedResult: Double?
    let growthPoints: [GrowthPoint]?
    
    // Metadata
    let notes: String?
    let tags: [String]
    
    init(id: UUID = UUID(), name: String, calculationType: CalculationMode, 
         createdDate: Date = Date(), modifiedDate: Date = Date(), projectId: UUID? = nil,
         initialInvestment: Double? = nil, outcomeAmount: Double? = nil, 
         timeInMonths: Double? = nil, irr: Double? = nil, 
         followOnInvestments: [FollowOnInvestment]? = nil,
         unitPrice: Double? = nil, successRate: Double? = nil, 
         outcomePerUnit: Double? = nil, investorShare: Double? = nil, 
         feePercentage: Double? = nil,
         calculatedResult: Double? = nil, growthPoints: [GrowthPoint]? = nil,
         notes: String? = nil, tags: [String] = []) throws {
        
        // Validate inputs during initialization
        try Self.validateName(name)
        try Self.validateCalculationInputs(
            calculationType: calculationType,
            initialInvestment: initialInvestment,
            outcomeAmount: outcomeAmount,
            timeInMonths: timeInMonths,
            irr: irr,
            unitPrice: unitPrice,
            successRate: successRate,
            outcomePerUnit: outcomePerUnit,
            investorShare: investorShare
        )
        
        if let followOns = followOnInvestments {
            try Self.validateFollowOnInvestments(followOns)
        }
        
        self.id = id
        self.name = name
        self.calculationType = calculationType
        self.createdDate = createdDate
        self.modifiedDate = modifiedDate
        self.projectId = projectId
        self.initialInvestment = initialInvestment
        self.outcomeAmount = outcomeAmount
        self.timeInMonths = timeInMonths
        self.irr = irr
        self.followOnInvestments = followOnInvestments
        self.unitPrice = unitPrice
        self.successRate = successRate
        self.outcomePerUnit = outcomePerUnit
        self.investorShare = investorShare
        self.feePercentage = feePercentage
        self.calculatedResult = calculatedResult
        self.growthPoints = growthPoints
        self.notes = notes
        self.tags = tags
    }
    
    // MARK: - Validation Methods
    
    /// Validates the entire calculation for data integrity
    func validate() throws {
        try Self.validateName(name)
        try Self.validateCalculationInputs(
            calculationType: calculationType,
            initialInvestment: initialInvestment,
            outcomeAmount: outcomeAmount,
            timeInMonths: timeInMonths,
            irr: irr,
            unitPrice: unitPrice,
            successRate: successRate,
            outcomePerUnit: outcomePerUnit,
            investorShare: investorShare
        )
        
        if let followOns = followOnInvestments {
            try Self.validateFollowOnInvestments(followOns)
        }
    }
    
    /// Validates calculation name
    static func validateName(_ name: String) throws {
        guard !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw SavedCalculationValidationError.emptyName
        }
        
        guard name.count <= 100 else {
            throw SavedCalculationValidationError.invalidName("Name too long (max 100 characters)")
        }
        
        // Check for invalid characters
        let invalidChars = CharacterSet(charactersIn: "<>:\"/\\|?*")
        guard name.rangeOfCharacter(from: invalidChars) == nil else {
            throw SavedCalculationValidationError.invalidName("Contains invalid characters")
        }
    }
    
    /// Validates calculation inputs based on calculation type
    static func validateCalculationInputs(
        calculationType: CalculationMode,
        initialInvestment: Double?,
        outcomeAmount: Double?,
        timeInMonths: Double?,
        irr: Double?,
        unitPrice: Double? = nil,
        successRate: Double? = nil,
        outcomePerUnit: Double? = nil,
        investorShare: Double? = nil
    ) throws {
        var missingFields: [String] = []
        
        switch calculationType {
        case .calculateIRR:
            if let investment = initialInvestment {
                guard investment > 0 else { throw SavedCalculationValidationError.negativeInvestment }
            } else {
                missingFields.append("Initial Investment")
            }
            
            if let outcome = outcomeAmount {
                guard outcome > 0 else { throw SavedCalculationValidationError.negativeOutcome }
            } else {
                missingFields.append("Outcome Amount")
            }
            
            if let time = timeInMonths {
                guard time > 0 else { throw SavedCalculationValidationError.invalidTimeInMonths }
            } else {
                missingFields.append("Time in Months")
            }
            
        case .calculateOutcome:
            if let investment = initialInvestment {
                guard investment > 0 else { throw SavedCalculationValidationError.negativeInvestment }
            } else {
                missingFields.append("Initial Investment")
            }
            
            if let irrValue = irr {
                guard irrValue > -100 && irrValue < 1000 else { throw SavedCalculationValidationError.invalidIRR }
            } else {
                missingFields.append("IRR")
            }
            
            if let time = timeInMonths {
                guard time > 0 else { throw SavedCalculationValidationError.invalidTimeInMonths }
            } else {
                missingFields.append("Time in Months")
            }
            
        case .calculateInitial:
            if let outcome = outcomeAmount {
                guard outcome > 0 else { throw SavedCalculationValidationError.negativeOutcome }
            } else {
                missingFields.append("Outcome Amount")
            }
            
            if let irrValue = irr {
                guard irrValue > -100 && irrValue < 1000 else { throw SavedCalculationValidationError.invalidIRR }
            } else {
                missingFields.append("IRR")
            }
            
            if let time = timeInMonths {
                guard time > 0 else { throw SavedCalculationValidationError.invalidTimeInMonths }
            } else {
                missingFields.append("Time in Months")
            }
            
        case .calculateBlendedIRR:
            if let investment = initialInvestment {
                guard investment > 0 else { throw SavedCalculationValidationError.negativeInvestment }
            } else {
                missingFields.append("Initial Investment")
            }
            
            if let outcome = outcomeAmount {
                guard outcome > 0 else { throw SavedCalculationValidationError.negativeOutcome }
            } else {
                missingFields.append("Outcome Amount")
            }
            
            if let time = timeInMonths {
                guard time > 0 else { throw SavedCalculationValidationError.invalidTimeInMonths }
            } else {
                missingFields.append("Time in Months")
            }
            
        case .portfolioUnitInvestment:
            if let investment = initialInvestment {
                guard investment > 0 else { throw SavedCalculationValidationError.negativeInvestment }
            } else {
                missingFields.append("Investment Amount")
            }
            
            if let price = unitPrice {
                guard price > 0 else { throw SavedCalculationValidationError.negativeInvestment }
            } else {
                missingFields.append("Unit Price")
            }
            
            if let rate = successRate {
                guard rate >= 0 && rate <= 100 else { throw SavedCalculationValidationError.invalidIRR }
            } else {
                missingFields.append("Success Rate")
            }
            
            if let outcome = outcomePerUnit {
                guard outcome > 0 else { throw SavedCalculationValidationError.negativeOutcome }
            } else {
                missingFields.append("Outcome Per Unit")
            }
            
            if let share = investorShare {
                guard share >= 0 && share <= 100 else { throw SavedCalculationValidationError.invalidIRR }
            } else {
                missingFields.append("Investor Share")
            }
            
            if let time = timeInMonths {
                guard time > 0 else { throw SavedCalculationValidationError.invalidTimeInMonths }
            } else {
                missingFields.append("Time in Months")
            }
        }
        
        if !missingFields.isEmpty {
            throw SavedCalculationValidationError.missingRequiredFields(missingFields)
        }
    }
    
    /// Validates follow-on investments
    static func validateFollowOnInvestments(_ investments: [FollowOnInvestment]) throws {
        var errors: [String] = []
        
        for (index, investment) in investments.enumerated() {
            do {
                try investment.validate()
            } catch {
                errors.append("Investment \(index + 1): \(error.localizedDescription)")
            }
        }
        
        if !errors.isEmpty {
            throw SavedCalculationValidationError.invalidFollowOnInvestments(errors)
        }
    }
    
    /// Checks if calculation has all required data for its type
    var isComplete: Bool {
        do {
            try validate()
            return true
        } catch {
            return false
        }
    }
    
    /// Returns a summary of the calculation for display purposes
    var summary: String {
        switch calculationType {
        case .calculateIRR:
            if let result = calculatedResult {
                return "IRR: \(String(format: "%.2f", result))%"
            } else {
                return "IRR calculation"
            }
        case .calculateOutcome:
            if let result = calculatedResult {
                return "Outcome: $\(String(format: "%.2f", result))"
            } else {
                return "Outcome calculation"
            }
        case .calculateInitial:
            if let result = calculatedResult {
                return "Initial: $\(String(format: "%.2f", result))"
            } else {
                return "Initial investment calculation"
            }
        case .calculateBlendedIRR:
            if let result = calculatedResult {
                return "Blended IRR: \(String(format: "%.2f", result))%"
            } else {
                return "Blended IRR calculation"
            }
        case .portfolioUnitInvestment:
            if let result = calculatedResult {
                return "Portfolio IRR: \(String(format: "%.2f", result))%"
            } else {
                return "Portfolio unit investment"
            }
        }
    }
    
    /// Creates a copy with updated modification date
    func withUpdatedModificationDate() -> SavedCalculation {
        return try! SavedCalculation(
            id: id,
            name: name,
            calculationType: calculationType,
            createdDate: createdDate,
            modifiedDate: Date(),
            projectId: projectId,
            initialInvestment: initialInvestment,
            outcomeAmount: outcomeAmount,
            timeInMonths: timeInMonths,
            irr: irr,
            followOnInvestments: followOnInvestments,
            unitPrice: unitPrice,
            successRate: successRate,
            outcomePerUnit: outcomePerUnit,
            investorShare: investorShare,
            feePercentage: feePercentage,
            calculatedResult: calculatedResult,
            growthPoints: growthPoints,
            notes: notes,
            tags: tags
        )
    }
}

// MARK: - Serialization Extensions
extension SavedCalculation {
    /// Serializes follow-on investments to JSON data for Core Data storage
    var followOnInvestmentsData: Data? {
        guard let investments = followOnInvestments else { return nil }
        
        do {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            return try encoder.encode(investments)
        } catch {
            print("Failed to encode follow-on investments: \(error)")
            return nil
        }
    }
    
    /// Deserializes follow-on investments from JSON data
    static func followOnInvestments(from data: Data?) -> [FollowOnInvestment]? {
        guard let data = data else { return nil }
        
        do {
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            return try decoder.decode([FollowOnInvestment].self, from: data)
        } catch {
            print("Failed to decode follow-on investments: \(error)")
            return nil
        }
    }
    
    /// Serializes growth points to JSON data for Core Data storage
    var growthPointsData: Data? {
        guard let points = growthPoints else { return nil }
        
        do {
            let encoder = JSONEncoder()
            return try encoder.encode(points)
        } catch {
            print("Failed to encode growth points: \(error)")
            return nil
        }
    }
    
    /// Deserializes growth points from JSON data
    static func growthPoints(from data: Data?) -> [GrowthPoint]? {
        guard let data = data else { return nil }
        
        do {
            let decoder = JSONDecoder()
            return try decoder.decode([GrowthPoint].self, from: data)
        } catch {
            print("Failed to decode growth points: \(error)")
            return nil
        }
    }
    
    /// Serializes tags array to JSON string
    var tagsJSON: String {
        do {
            let encoder = JSONEncoder()
            let data = try encoder.encode(tags)
            return String(data: data, encoding: .utf8) ?? "[]"
        } catch {
            print("Failed to encode tags: \(error)")
            return "[]"
        }
    }
    
    /// Deserializes tags from JSON string
    static func tags(from json: String?) -> [String] {
        guard let json = json, let data = json.data(using: .utf8) else { return [] }
        
        do {
            let decoder = JSONDecoder()
            return try decoder.decode([String].self, from: data)
        } catch {
            print("Failed to decode tags: \(error)")
            return []
        }
    }
}

// MARK: - Project Validation Errors
enum ProjectValidationError: LocalizedError {
    case emptyName
    case invalidName(String)
    case invalidDescription(String)
    case invalidColor(String)
    
    var errorDescription: String? {
        switch self {
        case .emptyName:
            return "Project name cannot be empty"
        case .invalidName(let reason):
            return "Invalid project name: \(reason)"
        case .invalidDescription(let reason):
            return "Invalid project description: \(reason)"
        case .invalidColor(let reason):
            return "Invalid project color: \(reason)"
        }
    }
}

struct Project: Identifiable, Codable, Equatable, Hashable {
    let id: UUID
    let name: String
    let description: String?
    let createdDate: Date
    let modifiedDate: Date
    let color: String? // Hex color for UI
    
    init(id: UUID = UUID(), name: String, description: String? = nil,
         createdDate: Date = Date(), modifiedDate: Date = Date(), color: String? = nil) throws {
        
        // Validate inputs during initialization
        try Self.validateName(name)
        if let desc = description {
            try Self.validateDescription(desc)
        }
        if let colorValue = color {
            try Self.validateColor(colorValue)
        }
        
        self.id = id
        self.name = name
        self.description = description
        self.createdDate = createdDate
        self.modifiedDate = modifiedDate
        self.color = color
    }
    
    // MARK: - Validation Methods
    
    /// Validates the entire project for data integrity
    func validate() throws {
        try Self.validateName(name)
        if let desc = description {
            try Self.validateDescription(desc)
        }
        if let colorValue = color {
            try Self.validateColor(colorValue)
        }
    }
    
    /// Validates project name
    static func validateName(_ name: String) throws {
        guard !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw ProjectValidationError.emptyName
        }
        
        guard name.count <= 50 else {
            throw ProjectValidationError.invalidName("Name too long (max 50 characters)")
        }
        
        // Check for invalid characters
        let invalidChars = CharacterSet(charactersIn: "<>:\"/\\|?*")
        guard name.rangeOfCharacter(from: invalidChars) == nil else {
            throw ProjectValidationError.invalidName("Contains invalid characters")
        }
    }
    
    /// Validates project description
    static func validateDescription(_ description: String) throws {
        guard description.count <= 500 else {
            throw ProjectValidationError.invalidDescription("Description too long (max 500 characters)")
        }
    }
    
    /// Validates project color (hex format)
    static func validateColor(_ color: String) throws {
        let hexPattern = "^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$"
        let regex = try NSRegularExpression(pattern: hexPattern)
        let range = NSRange(location: 0, length: color.utf16.count)
        
        guard regex.firstMatch(in: color, options: [], range: range) != nil else {
            throw ProjectValidationError.invalidColor("Must be a valid hex color (e.g., #FF0000)")
        }
    }
    
    /// Checks if project has valid data
    var isValid: Bool {
        do {
            try validate()
            return true
        } catch {
            return false
        }
    }
    
    /// Creates a copy with updated modification date
    func withUpdatedModificationDate() -> Project {
        return try! Project(
            id: id,
            name: name,
            description: description,
            createdDate: createdDate,
            modifiedDate: Date(),
            color: color
        )
    }
    
    /// Default project colors for UI
    static let defaultColors = [
        "#007AFF", // Blue
        "#34C759", // Green
        "#FF9500", // Orange
        "#FF3B30", // Red
        "#AF52DE", // Purple
        "#FF2D92", // Pink
        "#5AC8FA", // Light Blue
        "#FFCC00", // Yellow
        "#FF6B6B", // Light Red
        "#4ECDC4"  // Teal
    ]
}

// MARK: - Project-Calculation Relationship Extensions
extension Project {
    /// Calculates statistics for calculations associated with this project
    func calculateStatistics(from calculations: [SavedCalculation]) -> ProjectStatistics {
        let projectCalculations = calculations.filter { $0.projectId == self.id }
        
        return ProjectStatistics(
            totalCalculations: projectCalculations.count,
            completedCalculations: projectCalculations.filter { $0.isComplete }.count,
            lastModified: projectCalculations.map { $0.modifiedDate }.max(),
            calculationTypes: Dictionary(grouping: projectCalculations, by: { $0.calculationType })
                .mapValues { $0.count }
        )
    }
}

struct ProjectStatistics {
    let totalCalculations: Int
    let completedCalculations: Int
    let lastModified: Date?
    let calculationTypes: [CalculationMode: Int]
    
    var completionRate: Double {
        guard totalCalculations > 0 else { return 0.0 }
        return Double(completedCalculations) / Double(totalCalculations)
    }
}
